src/Mvc/Mvc.Razor/src/HelperResult.cs

```diff
@@ -24,13 +24,21 @@ public HelperResult(Func<TextWriter, Task> asyncAction)
     {
         ArgumentNullException.ThrowIfNull(asyncAction);
 
-        _asyncAction = asyncAction;
+        // Hopefully people don't really do async work in the callback, but previously we were just blocking in WriteTo instead.
+        using var stringWriter = new StringWriter();
+        asyncAction(stringWriter).GetAwaiter().GetResult();
+        _content = stringWriter.ToString();
```

Comment:
This is still problematic, isn't it?

With the previous version of the code, if I have a loop where I do something async that takes (let's say 1s), even if I blocked later on, I would at worst wait 1s, as all the functions will be fired without waiting.

With this change, it means that all the calls will be sequenced, and I will end up waiting 100s instead of 1s.

I question what's the value of fixing a somewhat obscure scenario (placing an `if` inside an attribute, inside a loop) over potentially introducing a regression for existing apps here. I would like us to analyze and determine where the HelperResult is used by the compiler before we decide to do something.

Reply:
> With the previous version of the code, if I have a loop where I do something async that takes (let's say 1s), even if I blocked later on, I would at worst wait 1s, as all the functions will be fired without waiting.

These calls would have never run in parallel. Before this change, they were still sequenced because they were being run inside of blocking calls to `HelperResult.WriteTo()` here inside this loop in `ViewBuffer.WriteToAsync()`:

https://github.com/dotnet/aspnetcore/blob/16907a1efa1b53ea0c91984f90d8cef314fb4420/src/Mvc/Mvc.ViewFeatures/src/Buffers/ViewBuffer.cs#L211-L250

I have no Idea why `HelperResult` was never updated to implement `IHtmlAsyncContent`. It definitely looks like it should, but this is not a regression. It blocks just as much as before, just earlier.

Reply:
@halter73 For this particular case you are right, I believe. However, I think it's the case that IHtmlContent instances get "moved" around, so there can be cases where the content is not evaluated up until WriteTo is called, which might happen at a later stage. My point was that the type is meant to represent a "deferred write operation" and by inlining the evaluation it's no longer that.

The tricky thing that worries me here is that we are unlikely to see an app break, but we will potentially see many apps experience different behaviors, which will be hard to explain.

The Razor view engine has been synchronous for a long time (just the writing part), and we considered it a significant (and breaking) task to make it async. In general, the process for "writing to the buffer". In general, it is expected that for MVC and Razor Pages, you have already computed the data you need to render the view ahead of time and you don't do async work during the rendering process itself. Its similar in Blazor, you can't `await` inside the body of a component render.

Hope this helps add some context.